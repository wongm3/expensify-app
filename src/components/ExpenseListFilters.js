import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters';
import { DateRangePicker } from 'react-dates';

class ExpenseListFilters extends React.Component {

  state = {
    calendarFocused: null
  };

  onDatesChange = ({ startDate, endDate }) => {
    this.props.dispatch(setStartDate(startDate));
    this.props.dispatch(setEndDate(endDate));
  };

  onFocusChange = (calendarFocused) => {
    this.setState(() => ({ calendarFocused }));
  };

  render() {
    const { filters, dispatch } = this.props;
    return (
      <div>
        <input type="text" value={filters.text} onChange={(e) => {
          dispatch(setTextFilter(e.target.value));
        }}/>
        <select value={filters.sortBy} onChange={(e) => {
          if (e.target.value === 'amount') {
            dispatch(sortByAmount());
          } else if (e.target.value === 'date') {
            dispatch(sortByDate());
          }
        }}>
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
        <DateRangePicker
          startDate={filters.startDate}
          endDate={filters.endDate}
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          showClearDates={true}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
      </div>
    );
  }
}

const mapStateToProps = ({ filters }) => {
  return {
    filters
  }
};

export default connect(mapStateToProps)(ExpenseListFilters);
