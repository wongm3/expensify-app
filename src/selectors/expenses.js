import moment from 'moment';

// Get visible expenses

export const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate}) => {
  return expenses.filter(expense => {
    const createdAtMoment = moment(expense.createdAt);
    const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
    const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
    const textMatch = expense.description.toUpperCase().includes(text.toUpperCase());

    return startDateMatch && endDateMatch && textMatch;
  }).sort((a, b) => {
    if (sortBy === 'date') {
      return a.createdAt < b.createdAt ? 1 : -1;
    } else if (sortBy === 'amount') {
      return a.amount < b.amount ? 1 : -1;
    }
  });
};