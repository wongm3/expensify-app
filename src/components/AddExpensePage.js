import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startAddExpense } from '../actions/expenses';

const AddExpensePage = ({ startAddExpense, history }) => (
  <div>
    <h1>Add Expense</h1>
    <ExpenseForm
      onSubmit={(expense) => {
        startAddExpense(expense);
        history.push('/');
      }}
    />
  </div>
);

const mapDispatchToProps = (dispatch) => ({
  startAddExpense: (expense) => dispatch(startAddExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);