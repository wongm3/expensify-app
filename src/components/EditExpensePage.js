import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense } from '../actions/expenses';

const EditExpensePage = ({ match: { params: { id } }, expense, dispatch, history }) => (
  <div>
    <h1>Edit Expense</h1>
    <ExpenseForm
      expense={expense}
      onSubmit={(expense) => {
        dispatch(editExpense(id, expense));
        history.push('/');
      }}
    />
    <button onClick={() => {
      dispatch(removeExpense({ id }));
      history.push('/');
    }}>Remove</button>
  </div>
);

const mapStateToProps = ({ expenses }, { match: { params: { id } } }) => {
  return {
    expense: expenses.find(expense => expense.id === id)
  }
};

export default connect(mapStateToProps)(EditExpensePage);