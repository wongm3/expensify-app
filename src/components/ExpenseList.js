import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import { getVisibleExpenses } from '../selectors/expenses';

export const ExpenseList = ({ expenses }) => {
  return (
    <div>
      <h1>Expense List</h1>
      {
        expenses.length === 0 ? (
          <p>No expenses</p>
        ) : (
          expenses.map((expense) => <ExpenseListItem key={expense.id} {...expense} />)
        )
      }
    </div>
  );
};

const mapStateToProps = ({ expenses, filters }) => {
  return {
    expenses: getVisibleExpenses(expenses, filters)
  };
};

export default connect(mapStateToProps)(ExpenseList);
