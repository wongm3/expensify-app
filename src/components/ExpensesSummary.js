import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import { getVisibleExpenses, getExpensesTotal } from '../selectors/expenses';

export const ExpensesSummary = ({ expensesCount, expensesTotal }) => (
  <div>
    Viewing { expensesCount } expense{ expensesCount !== 1 && 's' } totalling { numeral(expensesTotal / 100).format('$0,0.00') }
  </div>
);

const mapStateToProps = ({ expenses, filters }) => ({
  expensesCount: getVisibleExpenses(expenses, filters).length,
  expensesTotal: getExpensesTotal(getVisibleExpenses(expenses, filters))
});

export default connect(mapStateToProps)(ExpensesSummary);