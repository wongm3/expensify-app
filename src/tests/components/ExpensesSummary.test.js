import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';
import expenses from '../fixtures/expenses';
import { getExpensesTotal } from '../../selectors/expenses';

test('should render ExpensesSummary with an expense correctly', () => {
  const wrapper = shallow(<ExpensesSummary expenseCount={1} expensesTotal={getExpensesTotal([expenses[0]])} />);

  expect(wrapper).toMatchSnapshot();
});

test('should render ExpensesSummary with multiple expenses correctly', () => {
  const wrapper = shallow(<ExpensesSummary expenseCount={expenses.length} expensesTotal={getExpensesTotal(expenses)} />);

  expect(wrapper).toMatchSnapshot();
});