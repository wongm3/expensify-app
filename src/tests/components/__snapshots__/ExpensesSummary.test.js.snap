// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpensesSummary with an expense correctly 1`] = `
<div>
  Viewing 
   expense
  s
   totalling 
  $1.95
</div>
`;

exports[`should render ExpensesSummary with multiple expenses correctly 1`] = `
<div>
  Viewing 
   expense
  s
   totalling 
  $1,141.95
</div>
`;
